irmaseal-core/src/stream/sealer.rs

```diff
@@ -1,88 +1,92 @@
-use ctr::stream_cipher::{NewStreamCipher, StreamCipher};
-use hmac::Mac;
-use rand::{CryptoRng, Rng};
 use core::convert::TryFrom;
+use futures::{AsyncReadExt, AsyncWriteExt};
+use hmac::Mac;
 use postcard::to_slice;
+use rand::{CryptoRng, Rng};
 
 use crate::stream::*;
+use crate::Error::{ReadError, WriteError};
 use crate::*;
 
 /// Sealer for an bytestream, which converts it into an IRMAseal encrypted bytestream.
-pub struct Sealer<'a, W: Writable> {
+pub struct Sealer<W: AsyncWrite + Unpin> {
     aes: SymCrypt,
     hmac: Verifier,
-    w: &'a mut W,
+    output_writer: W,
 }
 
-impl<'a, W: Writable> Sealer<'a, W> {
-    pub fn new<R: Rng + CryptoRng>(
+impl<W: AsyncWrite + Unpin> Sealer<W> {
+    pub async fn new<R: Rng + CryptoRng>(
         i: Identity,
         pk: &PublicKey,
         rng: &mut R,
-        w: &'a mut W,
-    ) -> Result<Sealer<'a, W>, Error> {
+        mut w: W,
+    ) -> Result<Self, Error> {
         let derived = i.derive()?;
         let (c, k) = ibe::kiltz_vahlis_one::encrypt(&pk.0, &derived, rng);
 
         let (aeskey, mackey) = crate::stream::util::derive_keys(&k);
         let iv = crate::stream::util::generate_iv(rng);
```

Comment:
Since the IV is in the metadata now, my previous code structure is a bit broken. My idea first was to initate a `Sealer` with an identity and then `seal` can be called with an input and output stream to encrypt the input stream using that identity.

This is however not really possible anymore, since now that would imply that the same IV is used for multiple byte streams. That is not acceptable. @rowanG077 is it easy to generate the IV on a later moment or is it a fixed part of the metadata now? When the latter is the case, I should refactor it in such a way that generating the metadata bytes is done in the seal function. Otherwise I can precompute some of the metadata here and add the IV in the `seal` method.

Reply:
In my intial idea both the `r` (`AsyncRead`) and the `w` (`AsyncWrite`) where parameters of seal. Now `w` is a parameter of `new` only to get it working for now. This should change depending on the approach we choose.

Reply:
I now changed the signature of `seal` such that it uses `mut self` instead of `&mut self`, so now it will consume the `Sealer` object and therefore it will not be possible anymore to call `seal` multiple times.